flower/utils/results/result.py

```diff
@@ -0,0 +1,115 @@
+        the field names that Celery stores when persisting a result object to the backend.
+        """
+        self.task_id = task_id
+        self.status = status
+        self.date_done: datetime.datetime = dateutil.parser.parse(date_done)
+        self.result = result
+        self.traceback = traceback
+        self.name = name
+        self.args = args
+        self.kwargs = kwargs
+
+    @property
+    def result_extended(self) -> bool:
+        """
+        :return: Value of the Celery' app's `result_extended` setting when this Result was persisted to the backend. If
+          this is False, then some of the fields like name, args, kwargs will be null due to Celery storing only limited
+          fields in the backend.
+        """
+        return self.name is not None
+
+    def to_render_dict(self) -> dict[str, Any]:
+        """
+        Convert this data class record into an object that can be formatted by a template
+        """
+        return {
+            "result_extended": self.result_extended,
+            "name": self.name,
+            "task_id": self.task_id,
+            "status": self.status,
+            "date_done": self.date_done.timestamp(),
+            "args": repr(self.args),
+            "kwargs": repr(self.kwargs),
+            "result": repr(self.result),
+            "traceback": str(self.traceback),
+        }
+
+
+ResultIdWithResultPair: TypeAlias = tuple[str, Result]
+
+
+@total_ordering
+class ResultHeapEntry:
+    """
+    TODO: document
+    """
+    def __init__(self, result: Result, *, reverse_ordering: bool):
+        self.result = result
+        self.reverse_ordering = reverse_ordering
+        self.timetup = kombu.clocks.timetuple(None, result.date_done.timestamp(), result.task_id, result)
+
+    def __lt__(self, other: "ResultHeapEntry") -> bool:
+        if self.reverse_ordering:
+            return self.timetup > other.timetup
+        else:
+            return self.timetup < other.timetup
+
+    def __eq__(self, other: "ResultHeapEntry") -> bool:
+        return self.timetup == other.timetup
+
+
+class ResultHeap(Iterable[ResultIdWithResultPair]):
+    """
+    TODO: document
+    """
+    def __init__(self, *, heap_size_limit: int, reverse_ordering: bool):
+        self._heap_size_limit = heap_size_limit
+        self._reverse_ordering = reverse_ordering
+        self._heap: list[ResultHeapEntry] = []
+
+    def push(self, result: Result) -> Result | None:
+        """
+        TODO: document
+        """
+        new_heap_entry = ResultHeapEntry(result, reverse_ordering=self._reverse_ordering)
+
+        if len(self._heap) >= self._heap_size_limit:
+            popped_heap_entry: ResultHeapEntry = heapq.heappushpop(self._heap, new_heap_entry)
+            return popped_heap_entry.result
+
+        heapq.heappush(self._heap, new_heap_entry)
+        return None
+
+    def __iter__(self) -> Iterator[ResultIdWithResultPair]:
+        heap_copy = list(self._heap)
```

Comment:
i'm really curious if there are technical reasons to avoid the use of `copy.copy()`, dyk?

Reply:
I don't think there are any reasons to avoid. Pretty sure that in this case, `copy.copy(...)` and `list(...)` achieve exactly the same result since both of them do a shallow copy, and a shallow copy is what I intend here.

Reply:
seems fine in bench

```
dylang@ip-10-70-145-116():~$ python3 -m timeit --setup 'import random;a1 = [random.random() for i in range(1000)]; import copy' 'a2 = copy.copy(a1)'
100000 loops, best of 5: 2.96 usec per loop
dylang@ip-10-70-145-116():~$ python3 -m timeit --setup 'import random;a1 = [random.random() for i in range(1000)]; import copy' 'a2 = list(a1)'
100000 loops, best of 5: 2.62 usec per loop
```
shipit